strongbox-user-management/src/main/java/org/carlspring/strongbox/authentication/registry/AuthenticatorsRegistry.java

```diff
@@ -0,0 +1,136 @@
+package org.carlspring.strongbox.authentication.registry;
+
+import org.carlspring.strongbox.authentication.api.Authenticator;
+import org.carlspring.strongbox.authentication.registry.support.AuthenticatorsRegistrySerializer;
+
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.fasterxml.jackson.databind.module.SimpleModule;
+import com.google.common.collect.ImmutableList;
+
+/**
+ * @author Przemyslaw Fusik
+ */
+@JsonSerialize(using = AuthenticatorsRegistrySerializer.class)
+public class AuthenticatorsRegistry
+        implements Supplier<List<Authenticator>>
+{
+
+    final transient ReentrantLock lock = new ReentrantLock();
+
+    private final List<Authenticator> authenticators;
+
+    private List<Authenticator> lastSnapshot;
+
+    @Inject
+    private Optional<ObjectMapper> objectMapper;
+
+    public AuthenticatorsRegistry()
+    {
+        this(Collections.emptyList());
+    }
+
+
+    public AuthenticatorsRegistry(List<Authenticator> authenticators)
+    {
+        this.authenticators = new ArrayList<>(authenticators);
+        updateSnapshot();
+    }
+
+    /**
+     * Reloads the registry by replacing all authenticators using
+     * given collection.
+     *
+     * @param authenticators new collection of authenticators
+     */
+    public void reload(Collection<Authenticator> authenticators)
+    {
+        Objects.requireNonNull(authenticators, () -> "Required non-null replacing authenticators.");
+        lock.lock();
+        try
+        {
+            updateSnapshot();
+            this.authenticators.clear();
+            this.authenticators.addAll(authenticators);
+        }
+        finally
+        {
+            updateSnapshot();
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Reorders elements in the registry.
+     */
+    public void reorder(int first,
+                        int second)
+    {
+        if (first == second)
+        {
+            return;
+        }
+        lock.lock();
```

Comment:
I believe my solution meets the [rule](https://dev.carlspring.org/sonar/coding_rules#rule_key=findbugs%3AUL_UNRELEASED_LOCK_EXCEPTION_PATH) ... False positive ? Maybe someone else could have a look on it ?